mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+        )
+        final_action_state, metrics = scanned_act(action_state)
+
+        # Sample and learn
+        train_state = TrainState(
+            final_action_state.buffer_state,
+            params,
+            opt_state,
+            train_steps,
+            train_key,
+        )
+        final_train_state, losses = scanned_train(train_state)
+
+        next_learner_state = LearnerState(
+            final_action_state.obs,
+            final_action_state.terminal,
+            final_action_state.term_or_trunc,
+            final_action_state.action_selection_state.hidden_state,
+            final_action_state.env_state,
+            final_action_state.action_selection_state.time_steps,
+            final_train_state.train_steps,
+            final_train_state.opt_state,
+            final_action_state.buffer_state,
+            final_train_state.params,
+            new_key,
+        )
+
+        return next_learner_state, (metrics, losses)
+
+    pmaped_update_step = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(update_step, state, None, length=cfg.system.scan_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+
+    return pmaped_update_step
+
+
+def run_experiment(cfg: DictConfig) -> float:
+    cfg.arch.n_devices = len(jax.devices())
+    cfg = check_total_timesteps(cfg)
+
+    # Number of env steps before evaluating/logging.
+    steps_per_rollout = int(cfg.system.total_timesteps // cfg.arch.num_evaluation)
+    # Multiplier for a single env/learn step in an anakin system
+    anakin_steps = cfg.arch.n_devices * cfg.system.update_batch_size
+    # Number of env steps in one anakin style update.
+    anakin_act_steps = anakin_steps * cfg.arch.num_envs * cfg.system.rollout_length
+    # Number of steps to do in the scanned update method (how many anakin steps).
+    cfg.system.scan_steps = int(steps_per_rollout / anakin_act_steps)
+
+    pprint(OmegaConf.to_container(cfg, resolve=True))
+
+    # Initialise system and make learning/evaluation functions
+    (env, eval_env), q_net, q_mixer, opts, rb, learner_state, logger, key = init(cfg)
+    update = make_update_fns(cfg, env, q_net, q_mixer, opts, rb)
+
+    cfg.system.num_agents = env.num_agents
+
+    key, eval_key = jax.random.split(key)
+
+    def eval_act_fn(
+        params: FrozenDict, timestep: TimeStep, key: chex.PRNGKey, actor_state: ActorState
+    ) -> Tuple[chex.Array, ActorState]:
+        """The acting function that get's passed to the evaluator.
+        A custom function is needed for epsilon-greedy acting.
+        """
+        hidden_state = actor_state["hidden_state"]
+
+        term_or_trunc = timestep.last()
+        net_input = (timestep.observation, term_or_trunc[..., jnp.newaxis])
+        net_input = tree.map(lambda x: x[jnp.newaxis], net_input)  # add batch dim to obs
+        next_hidden_state, eps_greedy_dist = q_net.apply(params, hidden_state, net_input)
```

Comment:
You don't put 0 in here, but it's there in IQL, can you remove it from IQL :pray: